Crea un nuevo notebook de jupyter, guárdelo en la carpeta notebooks con el nombre 3_nb_***.ipynb donde *** es el github del integrante 1. Una vez que tenga todo el ejercicio listo, limpie los outputs, reinicie el kernel y corra todo el notebook. Luego, exporta el notebook como html y guárdalo en la carpeta output.


En el notebook 3_nb_***.ipynb, cargue los dos archivos d2019.csv y d2020.txt en dos dataframes llamados df2019 y df2020 respectivamente. Luego, usando el método melt, genera un dataframe llamado df_sii donde estén incluidos los datos del 2019 y del 2020, que tenga las columnas year,mes, dia, valor, y guardalo como csv en la carpeta data/interim con el nombre df_sii.csv.

In [1]:
# IMPORTAR LIBRERÍAS

import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pylab as plt
%matplotlib inline
import re

### Importar Archivos

In [2]:

df2019=pd.read_csv("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/d2019.csv")

In [3]:
df2019.head(3)

Unnamed: 0  Día     Ene     Feb     Mar     Abr     May    Jun     Jul  \
0           0    1     NaN  657.81  651.79  678.53     NaN    NaN  679.15   
1           1    2  694.77     NaN     NaN  672.56  678.71    NaN  677.62   
2           2    3  697.09     NaN     NaN  670.05  683.31  709.8  681.30   

      Ago     Sep     Oct  Nov     Dic  year  
0  700.82     NaN  728.21  NaN     NaN  2019  
1  704.88  720.20  729.38  NaN  812.13  2019  
2     NaN  724.72  727.80  NaN  804.29  2019

In [4]:
df2020=pd.read_csv("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/d2020.txt", sep=" ", header= 0)


In [5]:
df2020.head(3)

Unnamed: 0  Día     Ene     Feb     Mar     Abr  May     Jun     Jul  \
0           0    1     NaN     NaN     NaN  852.03  NaN  806.32  821.23   
1           1    2  748.74     NaN  818.32  863.81  NaN  796.46  816.29   
2           2    3  754.16  799.11  817.45  861.89  NaN  782.86  803.98   

      Ago     Sep     Oct     Nov     Dic  year  
0     NaN  776.46  788.15     NaN  767.29  2020  
1     NaN  770.59  785.04  771.92  760.16  2020  
2  757.06  771.45     NaN  769.17  755.34  2020

### Luego, usando el método melt, genera un dataframe llamado df_sii donde estén incluidos los datos del 2019 y del 2020, que tenga las columnas year,mes, dia, valor, y guardalo como csv en la carpeta data/interim con el nombre df_sii.csv.

In [6]:
df1920=pd.concat([df2019,df2020], axis=0)
print(df1920)

    Unnamed: 0  Día     Ene     Feb     Mar     Abr     May     Jun     Jul  \
0            0    1     NaN  657.81  651.79  678.53     NaN     NaN  679.15   
1            1    2  694.77     NaN     NaN  672.56  678.71     NaN  677.62   
2            2    3  697.09     NaN     NaN  670.05  683.31  709.80  681.30   
3            3    4  697.64  655.39  656.57  667.55     NaN  704.59  680.48   
4            4    5     NaN  653.80  659.07  666.32     NaN  696.78  677.64   
..         ...  ...     ...     ...     ...     ...     ...     ...     ...   
26          26   27  776.52  811.10  836.05  859.95  802.10     NaN  768.63   
27          27   28  787.86  816.26     NaN  856.76  816.47     NaN  768.81   
28          28   29  788.13     NaN     NaN  850.41  812.74     NaN  767.82   
29          29   30  788.18     NaN  835.23  836.62     NaN  816.36  759.18   
30          30   31  797.96     NaN  846.30     NaN     NaN     NaN  754.45   

       Ago     Sep     Oct     Nov     Dic  year  


In [7]:
df1920_1=df1920.drop(columns=['Unnamed: 0'])
df1920_1.columns

Index(['Día', 'Ene', 'Feb', 'Mar', 'Abr', 'May', 'Jun', 'Jul', 'Ago', 'Sep',
       'Oct', 'Nov', 'Dic', 'year'],
      dtype='object')

In [8]:
df1920_1.head(3)

Día     Ene     Feb     Mar     Abr     May    Jun     Jul     Ago     Sep  \
0    1     NaN  657.81  651.79  678.53     NaN    NaN  679.15  700.82     NaN   
1    2  694.77     NaN     NaN  672.56  678.71    NaN  677.62  704.88  720.20   
2    3  697.09     NaN     NaN  670.05  683.31  709.8  681.30     NaN  724.72   

      Oct  Nov     Dic  year  
0  728.21  NaN     NaN  2019  
1  729.38  NaN  812.13  2019  
2  727.80  NaN  804.29  2019

In [9]:
#generando el nuevo Dataframe con Melt

df_1920_2 = pd.melt(df1920_1, id_vars=['year','Día'], value_vars=['Ene', 'Feb', 'Mar', 'Abr', 'May', 'Jun', 'Jul', 'Ago', 'Sep','Oct', 'Nov', 'Dic'], var_name = 'mes', value_name = 'valor')
df_1920_2.head(3)

year  Día  mes   valor
0  2019    1  Ene     NaN
1  2019    2  Ene  694.77
2  2019    3  Ene  697.09

In [10]:
df_sii= df_1920_2
df_sii.columns

Index(['year', 'Día', 'mes', 'valor'], dtype='object')

In [11]:
df_sii = df_sii.reindex(columns=['year','mes','Día','valor'])

In [12]:
print(df_sii)

     year  mes  Día   valor
0    2019  Ene    1     NaN
1    2019  Ene    2  694.77
2    2019  Ene    3  697.09
3    2019  Ene    4  697.64
4    2019  Ene    5     NaN
..    ...  ...  ...     ...
739  2020  Dic   27     NaN
740  2020  Dic   28  710.26
741  2020  Dic   29  710.64
742  2020  Dic   30  711.24
743  2020  Dic   31     NaN

[744 rows x 4 columns]


In [13]:
#Guardando el archivo df_sii.csv

df_sii.to_csv('df_sii.csv')

### Usando el método groupby en df_sii:

Agrupa por mes, con el argumento as_index = True y usando el método aggregate con el método mean, imprime el resultado. Agrupa por [year, mes], con el argumento as_index = False y usando el método agg con el método mean, imprime el resultado Compara ambos resultados. ¿Qué cambia al usar as_index = False? ¿Al agrupar por mes, con respecto a [year,mes]?

In [14]:
#Agrupando los datos por mes, index True

df_sii_mes= df_sii.groupby('mes', as_index= True).aggregate(['mean'])

In [15]:
print(df_sii_mes)

       year   Día       valor
       mean  mean        mean
mes                          
Abr  2019.5  16.0  760.389048
Ago  2019.5  16.0  749.182619
Dic  2019.5  16.0  752.561750
Ene  2019.5  16.0  724.854773
Feb  2019.5  16.0  726.342000
Jul  2019.5  16.0  735.394318
Jun  2019.5  16.0  744.299024
Mar  2019.5  16.0  755.523488
May  2019.5  16.0  753.659500
Nov  2019.5  16.0  769.540976
Oct  2019.5  16.0  753.867907
Sep  2019.5  16.0  748.036154


In [16]:
#Agrupando por año y mes, index False

df_sii_year_mes= df_sii.groupby(['year','mes'], as_index= False).aggregate(['mean'])

In [17]:
print(df_sii_year_mes)

           Día       valor
          mean        mean
year mes                  
2019 Abr  16.0  667.399048
     Ago  16.0  713.703333
     Dic  16.0  770.390500
     Ene  16.0  677.061818
     Feb  16.0  656.304500
     Jul  16.0  686.059545
     Jun  16.0  692.409000
     Mar  16.0  667.678571
     May  16.0  692.003810
     Nov  16.0  776.530000
     Oct  16.0  721.032273
     Sep  16.0  718.442222
2020 Abr  16.0  853.379048
     Ago  16.0  784.661905
     Dic  16.0  734.733000
     Ene  16.0  772.647727
     Feb  16.0  796.379500
     Jul  16.0  784.729091
     Jun  16.0  793.718095
     Mar  16.0  839.375455
     May  16.0  821.805263
     Nov  16.0  762.884762
     Oct  16.0  788.267143
     Sep  16.0  773.402381


### Respuesta

En df_sii_mes el mes representa el indice del df, se agrupan los datos por los meses y en el resto de los parametros los valores se promedian
en df_sii_year_mes se agrupa por año y mes  y se cra un indice multiple, el resto de los parametros que se promedian 

In [18]:
#Agrupando por mes index false

df_sii_mes_2= df_sii.groupby('mes', as_index= False).aggregate(['mean'])

In [19]:
print(df_sii_mes_2)

       year   Día       valor
       mean  mean        mean
mes                          
Abr  2019.5  16.0  760.389048
Ago  2019.5  16.0  749.182619
Dic  2019.5  16.0  752.561750
Ene  2019.5  16.0  724.854773
Feb  2019.5  16.0  726.342000
Jul  2019.5  16.0  735.394318
Jun  2019.5  16.0  744.299024
Mar  2019.5  16.0  755.523488
May  2019.5  16.0  753.659500
Nov  2019.5  16.0  769.540976
Oct  2019.5  16.0  753.867907
Sep  2019.5  16.0  748.036154


### ¿Qué cambia al usar as_index = False? ¿Al agrupar por mes, con respecto a [year,mes]?
que se agrupa por el mes y el index solo tiene un nivel, el mes. Todos los parametros que quedan son promediados

### ¿Qué podrías inferir de los missing values y los días hábiles? Úsalos para crear una nueva variable llamada dia_de_la_semana en df_sii con valores "Lunes", "Martes", "Miércoles", "Jueves", "Viernes", "Sábado", "Domingo".

In [20]:
#Al revisar el Dataframe aparecen días que no existen en el calendario y deben ser eliminados para poder poner el día de la semana
#de forma correcta

df_feb = df_sii[ (df_sii['mes'] =='Feb') & (df_sii['Día'] >= 29)]
df_ab = df_sii[ (df_sii['mes'] =='Abr') & (df_sii['Día'] >= 31)]
df_jun = df_sii[ (df_sii['mes'] =='Jun') & (df_sii['Día'] >= 31)]
df_sep = df_sii[ (df_sii['mes'] =='Sep') & (df_sii['Día'] >= 31)]
df_nov = df_sii[ (df_sii['mes'] =='Nov') & (df_sii['Día'] >= 31)]

In [21]:
#Eliminando las filas del dataframe de días que no existen en el calendario (creando df_sii_fil1)

df_sii_fil1= df_sii.drop(df_sii.index[[90, 91, 92, 122, 123, 216, 247, 340, 371, 526, 557, 650, 681]])

In [22]:
#Cambiando los nombres de los meses del año por el número correspondiente al mes en df_sii(df_sii_fil2)

df_sii_fil2= df_sii_fil1.replace(['Ene','Feb','Mar','Abr','May','Jun','Jul','Ago','Sep','Oct','Nov','Dic'],['1','2','3','4','5','6','7','8','9','10','11','12'])
print(len(df_sii_fil2))
df_sii_fil2.head(3)

731


year mes  Día   valor
0  2019   1    1     NaN
1  2019   1    2  694.77
2  2019   1    3  697.09

In [23]:
#Creando columna fecha en df-sii_fil2

df_sii_fil2['Dates'] = df_sii_fil2['year'].astype(str)+'-'+ df_sii_fil2['mes'].astype(str) +'-'+ df_sii_fil2['Día'].astype(str)
print(len(df_sii_fil2))
df_sii_fil2.head(3)

731


year mes  Día   valor     Dates
0  2019   1    1     NaN  2019-1-1
1  2019   1    2  694.77  2019-1-2
2  2019   1    3  697.09  2019-1-3

In [24]:
#Encontrando los días de la semana por la fecha

def date (Dates):
    for date in df_sii_fil2:
        date = pd.DataFrame({'inputDate': df_sii_fil2['Dates']}) 
        date['inputDate'] = pd.to_datetime(date['inputDate']) 
        date['dayOfWeek'] = date['inputDate'].dt.day_name() 
        
        return (date)
        
dias_de_la_semana_sii= date(df_sii_fil2.Dates)
len(dias_de_la_semana_sii)

731

In [25]:
#Creando columna de dia de la semana para df_sii

dias_sii=dias_de_la_semana_sii.rename(columns={'dayOfWeek':'dias_de_la_semana'})
dias_sii=dias_sii.replace(['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday', 'Sunday'],['Lunes',  'Martes',  'Miércoles', 
                   'Jueves',  'Viernes',  'Sábado',  'Domingo'])
len(dias_sii)

731

In [26]:
#Agregando la columna dia de la semana por fecha

lista_dias_sii= dias_sii['dias_de_la_semana'].tolist()

df_sii_fil2['dias_de_la_semana']= lista_dias_sii
df_sii_fil2.head(3)

year mes  Día   valor     Dates dias_de_la_semana
0  2019   1    1     NaN  2019-1-1            Martes
1  2019   1    2  694.77  2019-1-2         Miércoles
2  2019   1    3  697.09  2019-1-3            Jueves

In [27]:
df_sii_dias_final= df_sii_fil2.drop(columns=['Dates'])
df_sii_dias_final.head(3)

year mes  Día   valor dias_de_la_semana
0  2019   1    1     NaN            Martes
1  2019   1    2  694.77         Miércoles
2  2019   1    3  697.09            Jueves

In [28]:
df_sii_fil2.duplicated().value_counts()

False    731
dtype: int64

### Respuesta:

##### Que los missing values corresponden a días feriados o festivos.

Explora el archivo dolar_observado_bc.xlsx y su estructura.

In [29]:
df_dolar_observado_80=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='80')
df_dolar_observado_90=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='90')
df_dolar_observado_00=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='00')
df_dolar_observado_10=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='10')
df_dolar_observado_20=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='20')


Lee cada hoja por separado con read_excel, usando los argumentos skiprows=2, dtype={'Periodo':str }

In [30]:
df_dolar_observado_80=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='80',skiprows= 2, dtype={'Periodo':str })
df_dolar_observado_90=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='90',skiprows= 2, dtype={'Periodo':str })
df_dolar_observado_00=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='00',skiprows= 2, dtype={'Periodo':str })
df_dolar_observado_10=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='10',skiprows= 2, dtype={'Periodo':str })
df_dolar_observado_20=pd.read_excel("C:/Users/drasu/Documents/Analisis de Datos/Tarea_1_parte_3/dolar_observado_bc.xlsx", sheet_name='20',skiprows= 2, dtype={'Periodo':str })


In [31]:
df_dolar_observado_80.head(3)

Periodo  1.Dólar observado
0  1982-08-09 00:00:00              55.65
1  1982-08-10 00:00:00              62.79
2  1982-08-11 00:00:00              64.07

In [32]:
df_dolar_observado_80.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2702 entries, 0 to 2701
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Periodo            2702 non-null   object 
 1   1.Dólar observado  1848 non-null   float64
dtypes: float64(1), object(1)
memory usage: 42.3+ KB


In [33]:
df_dolar_observado_90.head(3)

Periodo  1.Dólar observado
0  1990-01-01 00:00:00                NaN
1  1990-01-02 00:00:00             297.37
2  1990-01-03 00:00:00             297.32

In [34]:
df_dolar_observado_00.head(3)

Periodo  1.Dólar observado
0  2000-01-01 00:00:00                NaN
1  2000-01-02 00:00:00                NaN
2  2000-01-03 00:00:00             530.07

In [35]:
df_dolar_observado_10.head(3)

Periodo  1.Dólar observado
0  2010-01-01 00:00:00                NaN
1  2010-01-02 00:00:00                NaN
2  2010-01-03 00:00:00                NaN

In [36]:
df_dolar_observado_20.head(3)

Periodo  1.Dólar observado
0  2020-01-01 00:00:00                NaN
1  2020-01-02 00:00:00             748.74
2  2020-01-03 00:00:00             754.16

### Crea un dataframe que una todos los dataframes en uno solo con las mismas columnas, y nombralo df_bc_all.

In [37]:
#Creando Dataframe df_bc_all

df_bc_all= pd.concat([df_dolar_observado_80, df_dolar_observado_90, df_dolar_observado_00, df_dolar_observado_10, df_dolar_observado_20])
df_bc_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14633 entries, 0 to 973
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Periodo            14633 non-null  object 
 1   1.Dólar observado  9997 non-null   float64
dtypes: float64(1), object(1)
memory usage: 343.0+ KB


In [38]:
# Separando Perido en fecha(Dates) y tiempo (timne)

df_bc_all['Dates'] = pd.to_datetime(df_bc_all['Periodo']).dt.date
df_bc_all['Time'] = pd.to_datetime(df_bc_all['Periodo']).dt.time
df_bc_all.head(3)

Periodo  1.Dólar observado       Dates      Time
0  1982-08-09 00:00:00              55.65  1982-08-09  00:00:00
1  1982-08-10 00:00:00              62.79  1982-08-10  00:00:00
2  1982-08-11 00:00:00              64.07  1982-08-11  00:00:00

In [39]:
len(df_bc_all)

14633

In [40]:
# Creando los días de la semana a partir de la fecha

def date (Dates):
    for date in df_bc_all:
        date = pd.DataFrame({'inputDate':df_bc_all['Dates']}) 
        date['inputDate'] = pd.to_datetime(date['inputDate']) 
        date['dayOfWeek'] = date['inputDate'].dt.day_name() 
        return (date)

    
dias_de_la_semana= date(df_bc_all.Dates)
dias_de_la_semana.head(3)

inputDate  dayOfWeek
0 1982-08-09     Monday
1 1982-08-10    Tuesday
2 1982-08-11  Wednesday

In [41]:
#Reemplazando los días de la semana desde el ingles al español

dias_de_la_semana=dias_de_la_semana.replace(['Monday','Tuesday','Wednesday','Thursday','Friday','Saturday', 'Sunday'],['Lunes',  'Martes',  'Miércoles', 
                   'Jueves',  'Viernes',  'Sábado',  'Domingo'])
dias_de_la_semana.head(3)

inputDate  dayOfWeek
0 1982-08-09      Lunes
1 1982-08-10     Martes
2 1982-08-11  Miércoles

In [42]:
#separando la fecha por año, mes y dia en df_bc_all

df_bc_all ['year'] = pd.DatetimeIndex(df_bc_all ['Dates']). year
df_bc_all ['mes'] = pd.DatetimeIndex(df_bc_all['Dates']). month
df_bc_all['Día'] = pd.DatetimeIndex(df_bc_all['Dates']).day
df_bc_all.head(3)

Periodo  1.Dólar observado       Dates      Time  year  mes  \
0  1982-08-09 00:00:00              55.65  1982-08-09  00:00:00  1982    8   
1  1982-08-10 00:00:00              62.79  1982-08-10  00:00:00  1982    8   
2  1982-08-11 00:00:00              64.07  1982-08-11  00:00:00  1982    8   

   Día  
0    9  
1   10  
2   11

In [43]:
#Revisando el nombre de las columnas

df_bc_all.columns

Index(['Periodo', '1.Dólar observado', 'Dates', 'Time', 'year', 'mes', 'Día'], dtype='object')

In [44]:
#Eliminando las columnas que no utilizaremos

df_bc_all.drop(columns = ['Periodo', 'Dates', 'Time'])
         

1.Dólar observado  year  mes  Día
0                55.65  1982    8    9
1                62.79  1982    8   10
2                64.07  1982    8   11
3                60.52  1982    8   12
4                56.87  1982    8   13
..                 ...   ...  ...  ...
969                NaN  2022    8   27
970                NaN  2022    8   28
971             900.22  2022    8   29
972             890.90  2022    8   30
973             882.11  2022    8   31

[14633 rows x 4 columns]

In [45]:
#Ordenando las columnas del df_bc_all

df_bc_all = df_bc_all.reindex(columns=['year','mes','Día','1.Dólar observado'])
df_bc_all.head(3)

year  mes  Día  1.Dólar observado
0  1982    8    9              55.65
1  1982    8   10              62.79
2  1982    8   11              64.07

### Crea una carpeta dentro de la carpeta data/processed llamada dolar_bc_year, agrupa por year usando groupby y guarda un CSV para cada año desde 1982 hasta 2022, con los nombres dolar_bc_YYYY.csv donde YYYY es el año. Busca una manera de hacerlo iterando y no uno a uno.

In [46]:
#Agrupando el dataframe df_bc_all por año (df_year1)

df_year1=df_bc_all.groupby(['year'],as_index=True)
df_year1

In [47]:
#revisando las etiquetas del Dataframe df_bc_all
df_bc_all.keys()


Index(['year', 'mes', 'Día', '1.Dólar observado'], dtype='object')

In [48]:
#Separando el Dataframe df_bc_all por año, en varios Dataframes

resultados = {}  # Diccionario inicialmente vacío
df=pd.DataFrame() # Dataframe vacio

for year in range(1982, 2022):
    resultados[year] = df_bc_all[df_bc_all.year == year]
    for year, dataframe in resultados.items():
        df_year= dataframe
    print(df_year)
break
    

     year  mes  Día  1.Dólar observado
0    1982    8    9              55.65
1    1982    8   10              62.79
2    1982    8   11              64.07
3    1982    8   12              60.52
4    1982    8   13              56.87
..    ...  ...  ...                ...
140  1982   12   27              73.25
141  1982   12   28              73.52
142  1982   12   29              73.55
143  1982   12   30              73.57
144  1982   12   31                NaN

[145 rows x 4 columns]
     year  mes  Día  1.Dólar observado
145  1983    1    1                NaN
146  1983    1    2                NaN
147  1983    1    3              73.43
148  1983    1    4              73.83
149  1983    1    5              73.86
..    ...  ...  ...                ...
505  1983   12   27              87.42
506  1983   12   28              87.47
507  1983   12   29              87.50
508  1983   12   30              87.51
509  1983   12   31                NaN

[365 rows x 4 columns]
     year  mes  

SyntaxError: 'break' outside loop (2652282777.py, line 11)

In [49]:
#Guardando los Dataframes en archivos csv por año

resultados = {}  # Diccionario inicialmente vacío
df=pd.DataFrame()

for year in range(1982, 2023):
    resultados[year] = df_bc_all[df_bc_all.year == year]
    for year, dataframe in resultados.items():
        nombre_dataframe= f'dolar_bc_{year}.csv'
        dataframe.to_csv(nombre_dataframe)
    print(f'Generando el archivo {nombre_dataframe}')
    
break
    
  

Generando el archivo dolar_bc_1982.csv
Generando el archivo dolar_bc_1983.csv
Generando el archivo dolar_bc_1984.csv
Generando el archivo dolar_bc_1985.csv
Generando el archivo dolar_bc_1986.csv
Generando el archivo dolar_bc_1987.csv
Generando el archivo dolar_bc_1988.csv
Generando el archivo dolar_bc_1989.csv
Generando el archivo dolar_bc_1990.csv
Generando el archivo dolar_bc_1991.csv
Generando el archivo dolar_bc_1992.csv
Generando el archivo dolar_bc_1993.csv
Generando el archivo dolar_bc_1994.csv
Generando el archivo dolar_bc_1995.csv
Generando el archivo dolar_bc_1996.csv
Generando el archivo dolar_bc_1997.csv
Generando el archivo dolar_bc_1998.csv
Generando el archivo dolar_bc_1999.csv
Generando el archivo dolar_bc_2000.csv
Generando el archivo dolar_bc_2001.csv
Generando el archivo dolar_bc_2002.csv
Generando el archivo dolar_bc_2003.csv
Generando el archivo dolar_bc_2004.csv
Generando el archivo dolar_bc_2005.csv
Generando el archivo dolar_bc_2006.csv
Generando el archivo dola

SyntaxError: 'break' outside loop (1440702099.py, line 13)

### Crea un dataframe que tenga los años 2019 y 2020 que se llame df_bc_2019_2020

In [50]:
#Creando un Datframe desde los archivos creados 2019

dolar_bc_2019=pd.read_csv('dolar_bc_2019.csv')
print(len(dolar_bc_2019))
dolar_bc_2019.head(3)

365


Unnamed: 0  year  mes  Día  1.Dólar observado
0        3287  2019    1    1                NaN
1        3288  2019    1    2             694.77
2        3289  2019    1    3             697.09

In [51]:
#Creando un Datframe desde los archivos creados 2020

dolar_bc_2020=pd.read_csv('dolar_bc_2020.csv')
print(len(dolar_bc_2020))
dolar_bc_2020.head(3)

366


Unnamed: 0  year  mes  Día  1.Dólar observado
0           0  2020    1    1                NaN
1           1  2020    1    2             748.74
2           2  2020    1    3             754.16

In [52]:
#Creando el Dataframe df_bc_2019_2020

dolar_bc_2019=pd.read_csv('dolar_bc_2019.csv')
dolar_bc_2020=pd.read_csv('dolar_bc_2020.csv')
df_bc_2019_2020= pd.concat([dolar_bc_2019,dolar_bc_2020], axis=0)
print(len(df_bc_2019_2020))
df_bc_2019_2020.columns

731


Index(['Unnamed: 0', 'year', 'mes', 'Día', '1.Dólar observado'], dtype='object')

In [53]:
# Limpiendo el Dataframe, se eliminia la columna Unnamed

df_bc_2019_2020.drop(columns=['Unnamed: 0'])

year  mes  Día  1.Dólar observado
0    2019    1    1                NaN
1    2019    1    2             694.77
2    2019    1    3             697.09
3    2019    1    4             697.64
4    2019    1    5                NaN
..    ...  ...  ...                ...
361  2020   12   27                NaN
362  2020   12   28             710.26
363  2020   12   29             710.64
364  2020   12   30             711.24
365  2020   12   31                NaN

[731 rows x 4 columns]

### Compara los valores de df_bc_2019_2020 y df_sii. ¿Son iguales o diferentes?

In [54]:
#El dataframe df_sii filtrado, eliminando los valores que no existen por calendario es ahora llamado df_sii_fil2

print(len(df_sii_fil2))
df_sii_fil2.head(3)

731


year mes  Día   valor     Dates dias_de_la_semana
0  2019   1    1     NaN  2019-1-1            Martes
1  2019   1    2  694.77  2019-1-2         Miércoles
2  2019   1    3  697.09  2019-1-3            Jueves

In [55]:
# Uniformando los dataframes con las mismas columnas y nombres
#Transformando dataframe df_sii

df_sii_fil3=df_sii_fil2.drop(columns=['Dates','dias_de_la_semana'])
print(len(df_sii_fil3))
df_sii_fil3.head(3)

731


year mes  Día   valor
0  2019   1    1     NaN
1  2019   1    2  694.77
2  2019   1    3  697.09

In [56]:
#Transformando dataframe df_bc_2019_2020 (Se crea un nuevo dataframe llamdao: df_bc_2019_2020_m1)

df_bc_2019_2020_m= df_bc_2019_2020.drop(columns=['Unnamed: 0'])
df_bc_2019_2020_m1= df_bc_2019_2020_m.rename(columns={'1.Dólar observado': 'valor'})
print(len(df_bc_2019_2020_m1))
df_bc_2019_2020_m1.head(3)


731


year  mes  Día   valor
0  2019    1    1     NaN
1  2019    1    2  694.77
2  2019    1    3  697.09

In [57]:
#Uniformando los valores de los dataframes se cambia el formato del mes que era objeto a int

df_sii_fil3['mes']=df_sii_fil3['mes'].astype('int64')
print(type(df_sii_fil3))
df_sii_fil3.info()

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>
Int64Index: 731 entries, 0 to 743
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    731 non-null    int64  
 1   mes     731 non-null    int64  
 2   Día     731 non-null    int64  
 3   valor   499 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 28.6 KB


In [58]:
#Se revisa la informacion del dataframe df_bc_2019_2020_m1 para asegurar que sean comparables
print(type(df_bc_2019_2020_m1))
df_bc_2019_2020_m1.info()

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>
Int64Index: 731 entries, 0 to 365
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    731 non-null    int64  
 1   mes     731 non-null    int64  
 2   Día     731 non-null    int64  
 3   valor   499 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 28.6 KB


In [59]:
#Se revisa que tengan las mismas dimensiones

print(df_sii_fil3.shape)
print(df_bc_2019_2020_m1.shape)


(731, 4)
(731, 4)


In [60]:
#Utilizamos funcion para encontrar diferencias de las filas de los dataframes

def get_different_rows(source_df, new_df):
    """Retrona las filas del segundo dataframe que son diferentes al inicial"""
    merged_df = source_df.merge(new_df, indicator=True, how='outer')
    changed_rows_df = merged_df[merged_df['_merge'] == 'right_only']
    return changed_rows_df.drop('_merge', axis=1)

get_different_rows(df_sii_fil3, df_bc_2019_2020_m1)

Empty DataFrame
Columns: [year, mes, Día, valor]
Index: []

In [61]:
#Comparar por el metodo duplicate si los dataframes son diferentes

df_diff = pd.concat([df_sii_fil3, df_bc_2019_2020_m1]).drop_duplicates(keep= False)
df_diff.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    0 non-null      int64  
 1   mes     0 non-null      int64  
 2   Día     0 non-null      int64  
 3   valor   0 non-null      float64
dtypes: float64(1), int64(3)
memory usage: 0.0 bytes


### Compara los valores de df_bc_2019_2020 y df_sii. ¿Son iguales o diferentes?


#### Una vez que limpiamos los dataframes y comparamos los valores, podemos observar que son iguales.

### 4. Haz un merge de la columna dia_de_la_semana del df_sii al dataframe df_bc_all. 

In [62]:
#Visualizando Shape de Dataframes

print(df_sii.shape)
print(df_bc_all.shape)

(744, 4)
(14633, 4)


In [63]:
#Visualizando tipos de datos Dataframe Sii y uniformando para hacerlos compatibles

df_sii_dias_final['mes']=df_sii_dias_final['mes'].astype('int64')
df_sii_dias_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 731 entries, 0 to 743
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   year               731 non-null    int64  
 1   mes                731 non-null    int64  
 2   Día                731 non-null    int64  
 3   valor              499 non-null    float64
 4   dias_de_la_semana  731 non-null    object 
dtypes: float64(1), int64(3), object(1)
memory usage: 34.3+ KB


In [64]:
#Visualizando tipos de datos Dataframe df_bc_all y uniformando para hacerlos compatibles

df_bc_all.head(3)
df_bc_final= df_bc_all.rename(columns={'1.Dólar observado': 'valor'})
df_bc_final.head(3)
df_bc_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14633 entries, 0 to 973
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    14633 non-null  int64  
 1   mes     14633 non-null  int64  
 2   Día     14633 non-null  int64  
 3   valor   9997 non-null   float64
dtypes: float64(1), int64(3)
memory usage: 571.6 KB


In [65]:
# Uniendo los dataframes desde df_sii: se observa que Sii esté contenido en df_bc_all y queda como resultado el dataframe sii

df_final_m=pd.merge(df_sii_dias_final, df_bc_final)
df_final_m.head(3)

year  mes  Día   valor dias_de_la_semana
0  2019    1    1     NaN            Martes
1  2019    1    2  694.77         Miércoles
2  2019    1    3  697.09            Jueves

In [66]:
# Uniendo el dataframe df_bc_final con df_sii, usando Merge, se agrega la columna 'dias_de_la_semana' , cuyos valores que no existen en df_bc_final quedan con valores nulos

df_final=pd.merge( df_bc_final, df_sii_dias_final,  how='left', sort= False)
print(df_final.info())
df_final.head(3)


<class 'pandas.core.frame.DataFrame'>
Int64Index: 14633 entries, 0 to 14632
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   year               14633 non-null  int64  
 1   mes                14633 non-null  int64  
 2   Día                14633 non-null  int64  
 3   valor              9997 non-null   float64
 4   dias_de_la_semana  731 non-null    object 
dtypes: float64(1), int64(3), object(1)
memory usage: 685.9+ KB
None


year  mes  Día  valor dias_de_la_semana
0  1982    8    9  55.65               NaN
1  1982    8   10  62.79               NaN
2  1982    8   11  64.07               NaN

Los valores no nulos de días de la semana corresponden a los valores del df_sii

Para agregar los valores de los días de la semana se agrega la columna fecha



### 4.  Luego, extiende la variable dia_de_la_semana para todos los valores.

In [67]:
#Creando columna fecha en df_final

df_final['Dates'] = df_final['year'].astype(str)+'-'+ df_final['mes'].astype(str) +'-'+ df_final['Día'].astype(str)
print(len(df_final))
df_final.head(3)

14633


year  mes  Día  valor dias_de_la_semana      Dates
0  1982    8    9  55.65               NaN   1982-8-9
1  1982    8   10  62.79               NaN  1982-8-10
2  1982    8   11  64.07               NaN  1982-8-11

In [68]:
# Creando los días de la semana a partir de la fecha

def date (Dates):
    for date in df_final:
        date = pd.DataFrame({'Dates':df_final['Dates']}) 
        date['Dates'] = pd.to_datetime(date['Dates']) 
        date['dias_de_la_semana'] = date['Dates'].dt.day_name() 
    return (date)

dias= date(df_final.Dates)
dias.head(3)

Dates dias_de_la_semana
0 1982-08-09            Monday
1 1982-08-10           Tuesday
2 1982-08-11         Wednesday

In [69]:
#Agregando dias de la semana al dataframe

df_final['dias_de_la_semana_2']= dias['dias_de_la_semana']
print(df_final.head(3))

df_final_dias=df_final.drop(columns=['dias_de_la_semana'])
df_final_dias.head(3)

   year  mes  Día  valor dias_de_la_semana      Dates dias_de_la_semana_2
0  1982    8    9  55.65               NaN   1982-8-9              Monday
1  1982    8   10  62.79               NaN  1982-8-10             Tuesday
2  1982    8   11  64.07               NaN  1982-8-11           Wednesday


year  mes  Día  valor      Dates dias_de_la_semana_2
0  1982    8    9  55.65   1982-8-9              Monday
1  1982    8   10  62.79  1982-8-10             Tuesday
2  1982    8   11  64.07  1982-8-11           Wednesday

### Finalmente, crea una variable con las combinaciones [dia,mes] que tienen al menos un NaN que sea en día de semana (es decir, excluyendo Sábado y Domingo). 

In [70]:
#Creando Variable dia,mes

df_final_dias['dia , mes']= df_final_dias['Día'].astype(str) + ',' + df_final_dias['mes'].astype(str)
df_final_dias.head(3)

year  mes  Día  valor      Dates dias_de_la_semana_2 dia , mes
0  1982    8    9  55.65   1982-8-9              Monday       9,8
1  1982    8   10  62.79  1982-8-10             Tuesday      10,8
2  1982    8   11  64.07  1982-8-11           Wednesday      11,8

In [71]:
#creando columna para facilitar la busqueda

df_final_dias['dias_numero'] = df_final_dias.dias_de_la_semana_2.replace(['Monday','Tuesday','Wednesday', 'Thursday', 'Friday','Saturday','Sunday'], ['1', '2','3', '4', '5', '6','7' ])
df_final_dias.head(3)

year  mes  Día  valor      Dates dias_de_la_semana_2 dia , mes dias_numero
0  1982    8    9  55.65   1982-8-9              Monday       9,8           1
1  1982    8   10  62.79  1982-8-10             Tuesday      10,8           2
2  1982    8   11  64.07  1982-8-11           Wednesday      11,8           3

In [72]:
# Creando dataframe solo con valores nulos 

df_NaN= df_final_dias[df_final_dias['valor'].isnull()]
df_NaN.head(3)

year  mes  Día  valor      Dates dias_de_la_semana_2 dia , mes dias_numero
5   1982    8   14    NaN  1982-8-14            Saturday      14,8           6
6   1982    8   15    NaN  1982-8-15              Sunday      15,8           7
12  1982    8   21    NaN  1982-8-21            Saturday      21,8           6

In [73]:
df_NaN.columns


Index(['year', 'mes', 'Día', 'valor', 'Dates', 'dias_de_la_semana_2',
       'dia , mes', 'dias_numero'],
      dtype='object')

In [84]:
#Transformando tipo de dato para la busqueda

df_NaN['dias_numero']=df_NaN['dias_numero'].astype('int64')

#Creando Variable dia habil

dia_habil= df_NaN[df_NaN['dias_numero'] < 6]
dia_habil

C:\Users\drasu\AppData\Local\Temp\ipykernel_22808\1968062032.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_NaN['dias_numero']=df_NaN['dias_numero'].astype('int64')


year  mes  Día  valor       Dates dias_de_la_semana_2 dia , mes  \
64     1982   10   12    NaN  1982-10-12             Tuesday     12,10   
84     1982   11    1    NaN   1982-11-1              Monday      1,11   
121    1982   12    8    NaN   1982-12-8           Wednesday      8,12   
144    1982   12   31    NaN  1982-12-31              Friday     31,12   
158    1983    1   14    NaN   1983-1-14              Friday      14,1   
...     ...  ...  ...    ...         ...                 ...       ...   
14389  2021   12   31    NaN  2021-12-31              Friday     31,12   
14494  2022    4   15    NaN   2022-4-15              Friday      15,4   
14561  2022    6   21    NaN   2022-6-21             Tuesday      21,6   
14567  2022    6   27    NaN   2022-6-27              Monday      27,6   
14616  2022    8   15    NaN   2022-8-15              Monday      15,8   

       dias_numero  
64               2  
84               1  
121              3  
144              5  
158              5  
...            ...  
14389            5  
14494            5  
14561            2  
14567            1  
14616            1  

[456 rows x 8 columns]

In [103]:
dia_habil.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 456 entries, 64 to 14616
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   year                 456 non-null    int64  
 1   mes                  456 non-null    int64  
 2   Día                  456 non-null    int64  
 3   valor                0 non-null      float64
 4   Dates                456 non-null    object 
 5   dias_de_la_semana_2  456 non-null    object 
 6   dia , mes            456 non-null    object 
 7   dias_numero          456 non-null    int64  
dtypes: float64(1), int64(4), object(3)
memory usage: 32.1+ KB


In [85]:
dia_habil['dia , mes'].value_counts()

1,11     29
15,8     29
1,1      28
8,12     28
31,12    28
         ..
21,4      1
19,6      1
4,9       1
3,9       1
15,4      1
Name: dia , mes, Length: 78, dtype: int64

### 4.  Luego, cuenta cuantos NaN(incluyendo Sábados y Domingos) hay para cada combinación[dia/mes]. ¿Qué puedes inferir de los días con mayor cantidad de NaN?



In [104]:
dias_total=df_final_dias[df_final_dias['valor'].isnull()]
dias_total

year  mes  Día  valor      Dates dias_de_la_semana_2 dia , mes dias_numero
5   1982    8   14    NaN  1982-8-14            Saturday      14,8           6
6   1982    8   15    NaN  1982-8-15              Sunday      15,8           7
12  1982    8   21    NaN  1982-8-21            Saturday      21,8           6

In [108]:
dias_total.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4636 entries, 5 to 14629
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   year                 4636 non-null   int64  
 1   mes                  4636 non-null   int64  
 2   Día                  4636 non-null   int64  
 3   valor                0 non-null      float64
 4   Dates                4636 non-null   object 
 5   dias_de_la_semana_2  4636 non-null   object 
 6   dia , mes            4636 non-null   object 
 7   dias_numero          4636 non-null   object 
dtypes: float64(1), int64(3), object(4)
memory usage: 326.0+ KB


In [107]:
dias_total['dia , mes'].value_counts()

15,8     41
25,12    40
18,9     40
21,5     40
1,1      40
         ..
17,2     11
24,2     11
2,3      11
9,3      11
29,2      3
Name: dia , mes, Length: 366, dtype: int64

#### Se puede inferir comparando la cabtidad de fechas con NaN en días habiles total de 456 versus fechas con NaN en el total de los días 4636. Los días habiles representan un 9,8% del total de los NaN, de manera que los fines de semana deben estar vacios siempre.